server/src/main/java/org/elasticsearch/common/lucene/RegExp.java

```diff
@@ -0,0 +1,73 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.common.lucene;
+
+import org.apache.lucene.util.automaton.Automaton;
+import org.elasticsearch.core.SuppressForbidden;
+
+/**
+ * Simple wrapper for {@link org.apache.lucene.util.automaton.RegExp} that avoids throwing
+ * StackOverflowErrors when working with regular expressions as this will crash an Elasticsearch node.
+ * Instead, these StackOverflowErrors are turned into IllegalArgumentExceptions.
+ */
+@SuppressForbidden(reason = "catches StackOverflowError")
+public class RegExp {
+
+    private final org.apache.lucene.util.automaton.RegExp wrapped;
+
+    public RegExp(String s) {
+        try {
+            wrapped = new org.apache.lucene.util.automaton.RegExp(s);
+        } catch (StackOverflowError e) {
+            throw new IllegalArgumentException("failed to parse regexp", e);
+        }
+    }
+
+    public RegExp(String s, int syntax_flags) {
+        try {
+            wrapped = new org.apache.lucene.util.automaton.RegExp(s, syntax_flags);
+        } catch (StackOverflowError e) {
+            throw new IllegalArgumentException("failed to parse regexp", e);
+        }
+    }
+
+    public RegExp(String s, int syntax_flags, int match_flags) {
+        try {
+            wrapped = new org.apache.lucene.util.automaton.RegExp(s, syntax_flags, match_flags);
+        } catch (StackOverflowError e) {
+            throw new IllegalArgumentException("failed to parse regexp", e);
+        }
+    }
+
+    public Automaton toAutomaton() {
+        try {
+            return wrapped.toAutomaton();
+        } catch (StackOverflowError e) {
+            throw new IllegalArgumentException("failed to parse regexp", e);
+        }
+    }
+
+    public Automaton toAutomaton(int determinizeWorkLimit) {
+        try {
+            return wrapped.toAutomaton(determinizeWorkLimit);
+        } catch (StackOverflowError e) {
```

Comment:
I'm not sure what the exact concern was. One of the challenges with Errors (e.g. OOM) being thrown is that they can just happen anywhere in the code, and if the surrounding exception handling logic isn't handling and dealing with arbitrary Throwables, it can leave the program in an odd state (resources not being cleaned up etc). I'm not aware that it would do something bad with the JVM or bring it to an undefined state (but happy to learn more about that if it were the case). I therefore think that it's not a concern for Lucene's Regexp class, which isn't holding onto any resources or requires any kind of special clean-up. Maybe @rjernst has more thoughts on this?

Reply:
The Lucene regex class is indeed fairly contained. If there's any place safe to catch this then I think this is one of them.

Reply:
I talked with @ChrisHegarty about this. He [linked](https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/util/regex/Pattern.java#L1441) me to where the jdk does this when *it* compiles regexes. I'd say we're in good company then.

Reply:
But interestingly they don't propagate the whole overflowed stack when they rethrow, so I think it's worth making the change I suggested in my other comment.